### PR TITLE
[feature] Add "Linear Referencing" symbol layer type

### PR DESCRIPTION
**Pull request by nyalldawson**
This new symbol layer type allows placing text labels at regular intervals along a line (or at positions corresponding to existing vertices). Positions can be calculated using Cartesian distances, or interpolated from z/m values.

Functionality includes:

- Labels can be placed using fixed cartesian 2d distances, at regular linearly interpolated spacing calculated using the Z or M values in geometries, or at existing vertices
- Labels can show either the running total distance, or the linearly interpolated Z/M value
- Uses text renderer to draw labels, so the full range of functionality is available for the labels (including buffers, shadows, etc)
- Uses the QGIS numeric format classes to format numbers as strings, so users have full range of customisation options for eg decimal places
- An optional "skip multiples of" setting. If set, then labels which are a multiple of this value will be skipped over. This allows construction of complex referencing labels, eg where a symbol has two linear referencing symbol layers, one set to label every 100m in a small font, skipping multiples of 1000, and a second set to label every 1000m in a big bold font
- Labels are rendered using an angle calculated by averaging the linestring, so sharp tiny jaggies don't result in unslightly label rotation
- Optionally, markers can be placed at referenced points in the line string, using a full QGIS marker symbol (this allows eg showing a cross-hatch at the labeled point, for a "ruler" style line)
- Data defined control over the placement intervals, skip multiples setting, marker visibility and average angle calculation length

Notes:

- When using the distance-based placement or labels, the distances are calculated using 2D only, Cartesian calculations based on the original layer CRS. This could potentially be extended in future to expose options for 3D Cartesian distances, or ellipsoidal distance calculations.

Sponsored by the Swiss QGIS User Group

image

### CODE REVIEW

File: src/core/symbology/qgslinearreferencingsymbollayer.cpp [hunk 1]
```diff
@@ -0,0 +1,1003 @@
...
+  const double labelOffsetPainterUnitsX = context.renderContext().convertToPainterUnits( labelOffsetX, mLabelOffsetUnit, mLabelOffsetMapUnitScale );
+  const double labelOffsetPainterUnitsY = context.renderContext().convertToPainterUnits( labelOffsetY, mLabelOffsetUnit, mLabelOffsetMapUnitScale );
+  const double averageAngleDistancePainterUnits = context.renderContext().convertToPainterUnits( averageOver, mAverageAngleLengthUnit, mAverageAngleLengthMapUnitScale ) / 2;
+
+  // TODO (maybe?): if we don't have an original geometry, convert points to linestring and scale distance to painter units?
+  // in reality this line type makes no sense for rendering non-real feature geometries...
+  ( void )points;
+  const QgsAbstractGeometry *geometry = context.renderContext().geometry();
+  if ( !geometry )
+    return;
```
Comment (nirvn) [inline src/core/symbology/qgslinearreferencingsymbollayer.cpp:322]: Should we not return earlier to save cost of computing nbs above if we're not going to render anything?

File: src/core/symbology/qgslinearreferencingsymbollayer.cpp [hunk 2]
```diff
@@ -0,0 +1,1003 @@
...
+    if ( skipMultiples > 0 && qgsDoubleNear( std::fmod( currentDistance,  skipMultiples ), 0 ) )
+    {
+      prevX = thisX;
+      prevY = thisY;
+      continue;
+    }
+
+    const QPointF pt = pointToPainter( context, thisX, thisY, thisZ );
+
+    double calculatedAngle = 0;
```
Comment (nirvn) [inline src/core/symbology/qgslinearreferencingsymbollayer.cpp:806]: Can we move this down to the line that starts with calculatedAngle = std::fmod( QgsGeometryUtilsBase ... ? It's not used before then, it'd make the code reading easier.

File: src/core/symbology/qgslinearreferencingsymbollayer.h [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+
+    bool mShowMarker = false;
+    std::unique_ptr<QgsMarkerSymbol> mMarkerSymbol;
+
+    double mAverageAngleLength = 4;
+    Qgis::RenderUnit mAverageAngleLengthUnit = Qgis::RenderUnit::Millimeters;
+    QgsMapUnitScale mAverageAngleLengthMapUnitScale;
+
+    void renderGeometryPart( QgsSymbolRenderContext &context, const QgsAbstractGeometry *geometry, double labelOffsetPainterUnitsX, double labelOffsetPainterUnitsY, double skipMultiples, double averageAngleDistancePainterUnits, bool showMarker );
+    void renderLineString( QgsSymbolRenderContext &context, const QgsLineString *line, double labelOffsetPainterUnitsX, double labelOffsetPainterUnitsY, double skipMultiples, double averageAngleDistancePainterUnits, bool showMarker );
```
Comment (nirvn) [inline src/core/symbology/qgslinearreferencingsymbollayer.h:331]: Can we regroup the functions together in the private: section?

File: src/gui/numericformats/qgsnumericformatselectorwidget.h [hunk 4]
```diff
@@ -56,6 +57,13 @@ class GUI_EXPORT QgsNumericFormatSelectorWidget : public QgsPanelWidget, private
      */
     QgsNumericFormat *format() const SIP_TRANSFERBACK;
 
+    /**
+     * Register an expression context generator class that will be used to retrieve
+     * an expression context for the widget when required.
+     * \since QGIS 3.40
+     */
```
Comment (nirvn) [inline src/gui/numericformats/qgsnumericformatselectorwidget.h:64]: Note the ownership doesn't get passed on?

File: src/gui/symbology/qgssymbollayerwidget.cpp [hunk 5]
```diff
@@ -5458,3 +5460,243 @@ QgsSymbolLayer *QgsFilledLineSymbolLayerWidget::symbolLayer()
...
+      {
+        mLayer->setNumericFormat( widget->format() );
+        emit changed();
+      }
+    } );
+    panel->openPanel( widget );
+  }
+  else
+  {
+    // TODO!! dialog mode
```
Comment (nirvn) [inline src/gui/symbology/qgssymbollayerwidget.cpp:5700]: TODO?

File: src/core/symbology/qgslinearreferencingsymbollayer.cpp [hunk 6]
```diff
@@ -0,0 +1,1003 @@
...
+  res->setLabelOffsetMapUnitScale( mLabelOffsetMapUnitScale );
+  res->setShowMarker( mShowMarker );
+  res->setAverageAngleLength( mAverageAngleLength );
+  res->setAverageAngleUnit( mAverageAngleLengthUnit );
+  res->setAverageAngleMapUnitScale( mAverageAngleLengthMapUnitScale );
+
+  res->mTextFormat = mTextFormat;
+  res->mMarkerSymbol.reset( mMarkerSymbol ? mMarkerSymbol->clone() : nullptr );
+  if ( mNumericFormat )
+    res->mNumericFormat.reset( mNumericFormat->clone() );
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/core/symbology/qgslinearreferencingsymbollayer.cpp:125]: Cos the marker symbol is optional, whereas we expected that the numeric format should always be set. I'm playing things extra careful with the `if ( mNumericFormat )` check, but if somehow that happens then at least we won't overwrite the default format created for the cloned copy...

File: src/core/symbology/qgslinearreferencingsymbollayer.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/symbology/qgslinearreferencingsymbollayer.cpp:322]: Good point!

File: src/core/symbology/qgslinearreferencingsymbollayer.cpp (hunk 2 above)
Comment (nyalldawson) [reply src/core/symbology/qgslinearreferencingsymbollayer.cpp:806]: Done

**Comment by nyalldawson** [comment]
Thanks @nirvn, all addressed!

**Pull request merged**